jq255.md

```diff
@@ -0,0 +1,1425 @@
+
+```python
+def prepare_message_raw(msg):
+    return b'\x52' + msg
+
+def prepare_message_prehashed(hv, hashname):
+    return b'\x48' + hashname.encode() + b'\x00' + hv
+```
+
+The **recommended usage** is to apply jq255e and jq255s signatures on
+input messages pre-hashed with BLAKE2s with a 32-byte (256-bit) output.
+The signature scheme shall be designated, formally, as follows:
+
+  - **jq255e** and **jq255s**: signatures are computed over the
+    input message pre-hashed with BLAKE2s-256 (i.e. BLAKE2s with a
+    32-byte output).
+
+  - **jq255e-raw** and **jq255s-raw**: signatures are computed over
+    the raw input message, with no pre-hashing.
+
+  - **jq255e-hashname** and **jq255s-hashname**: signatures are computed
+    over the input message pre-hashed with the hash function whose
+    *symbolic name* is "hashname". Thus, "jq255e-sha256" means
+    "signature with the jq255e group, computed over an input message
+    pre-hashed with SHA-256".
+
+Thus, "jq255e", as a signature scheme name, is equivalent to
+"jq255e-blake2s".
+
+The following remarks apply:
+
+  - Regardless of the hash function used to pre-hash the message (if any),
+    the *internal* hash function used to compute the challenge is always
+    BLAKE2s. Within the scope of this specification, the use of BLAKE2s
+    internally is not configurable.
+
+  - Using a raw input conceptually provides additional protection
+    against collision attacks on weak hash functions. On the other hand,
+    raw inputs are not compatible with some kinds of streamed processing,
+    forcing either the sender or the recipient of a signed message to
+    buffer the entirety of the signed data. This is why pre-hashing with
+    a properly collision-resistant hash function is recommended in
+    general.
+
+#### Signature Generation
+
+A signature is computed with the following general process:
+
+  - A *per-signature nonce* `k` is produced. This is a secret scalar
+    that is chosen with uniform probability, and used only for a single
+    signature.
+
+  - A *commitment* is computed as the group element `R = k*G`.
+
+  - A *challenge* is computed by hashing together the commitment `R`,
+    the public key `Q`, and the prepared message `M`, yielding a
+    value `c`.
+
+  - The *response* is the scalar `s = k + c*sk` (interpreting `c` as
+    an integer which is then converted to a scalar).
+
+The signature value itself consists of `c` and an encoding of `s`.
+
+The signature generation process is described as the `sign()` function
+below, using the private key (`sk`), public key (`Q`) and prepared
+message (`M`) as input. That function calls the `generate_nonce()`
+function to generate `k` in a safe and deterministic way; that process
+uses an extra optional parameter called `seed`. More on this function
+is detailed later on.
+
+```python
+def generate_nonce(sk, Q, M, seed):
+    sh = hashlib.blake2s()
+    sh.update(private_key_encode(sk))
+    sh.update(public_key_encode(Q))
```

Comment:
Why include the verifying key/message in the nonce generation when those are committed as part of the challenge?

Reply:
Oh there's more at line 1163

Reply:
Injecting the message is necessary, otherwise two distinct messages might be signed with the same nonce (in the deterministic setup) and that would be Bad. Injecting the public key as well protects some weird "remote signer" implementations in which the caller can designate a private key and a public key independently, and confusion at that level can lead to some key recovery attacks (I know that EdDSA does not do that, and it has been pointed out in sternly worded tweets).

Reply:
OK; since this may, in the worst case of bad seed randomness, fall back to a full deterministicly generated nonce, I'm fine with this. // @FiloSottile